feat(STONEBLD-3431): Validate tasks and pipelines YAMLs

kyngs opened the pull request:
At the moment, errors in task and pipeline YAMLs only surface after the full CI run completes. This PR adds a GitHub Actions workflow that pre-validates task and pipeline YAMLs so that we can catch apply failures immediately.

This PR addresses [STONEBLD-3431](https://issues.redhat.com//browse/STONEBLD-3431) by adding a workflow to validate the tasks and pipelines YAMLs and whether they can be applied.

* Added a bash script to generate & apply tasks and pipelines via kubectl
* Added a GH Actions workflow to prepare a cluster via kind, and invoke the script
* Modified the gitops-pull-request.yaml so that it can be applied on its own (by defining the workspaces it uses)

chmeliik commented on .github/scripts/check_task_and_pipeline_yamls.sh:
```diff
@@ -0,0 +1,70 @@
+    for ignored in "${ignored_pipelines[@]}"; do
+      if [ "$ignored" == "$pipeline" ]; then
+        echo ">>> Ignoring pipeline: $pipeline"
+        continue 2
```
TIL about `continue [n]`, nice :taco:

kyngs commented on .github/scripts/check_task_and_pipeline_yamls.sh:
```diff
@@ -0,0 +1,70 @@
+          else
+            for yaml in "${version}"/*.yaml; do
+              if [ -f "$yaml" ] && [[ "$yaml" != *recipe.yaml ]]; then
+                to_apply+=("$yaml")
```
I wasn't aware of the convention. It does not apply to pipelines, and I didn't want to assume it does for tasks. Thanks for pointing that out, fixed.

kyngs replied:
> nd even tasks directories that have a kustomization.yaml should always contain the ${task_name}.yaml as well. We have other CI checks that ensure the task files are committed to the repo and kept up to date. So we don't really have to re-run kustomize here.

I think we have to re-run kustomize here. Each workflow is isolated, and the runs from previous ones won't get preserved here. Without running kustomize, applying the plain task fails.

kyngs commented on .github/scripts/check_task_and_pipeline_yamls.sh:
```diff
@@ -0,0 +1,70 @@
+            done
+          fi 
+          for file in "${to_apply[@]}"; do
+            kubectl apply -f "${file}"
```
Good catch! Using dry run in server mode will catch errors and not persist the changes to the cluster. Fixed.

kyngs commented on .github/scripts/check_task_and_pipeline_yamls.sh:
```diff
@@ -0,0 +1,70 @@
+    
+    for file in "${to_apply[@]}"; do
+      yq eval --inplace "del(.spec.tasks[].taskRef.version)" "${file}"
+      yq eval --inplace "del(.spec.finally[].taskRef.version)" "${file}"
```
Added the comment, it should be clear now.

kyngs commented:
> Nice job! 🎉
> 
> For demonstration purposes, could you temporarily add a commit that breaks the schema of some task/pipeline so that we can see what the failing check will look like?

@chmeliik Thanks for the review! I've addressed your comments and planted a bug to test the workflow out. It does catch the issue locally, but feel free to test it directly here on GitHub.

BTW, in case you didn't know, I suggest using `act` to test the workflow. Here's a one-liner for this PR:
```sh
act -W .github/workflows/check-task-and-pipeline-yamls.yaml pull_request -P ubuntu-24.04=ghcr.io/catthehacker/ubuntu:act-24.04
```

fmudry reviewed (commented):
Great job! :rocket:

chmeliik commented on .github/scripts/check_task_and_pipeline_yamls.sh:
```diff
@@ -0,0 +1,61 @@
+    done
+    echo ">>> Pipeline: $pipeline"
+    to_apply=()
+    if [ -e "$pipeline/kustomization.yaml" ]; then
+      kubectl kustomize "$pipeline" > "$pipeline/generated_pipeline.yaml"
+      to_apply+=("$pipeline/generated_pipeline.yaml")
+    else
+      for yaml in "$pipeline"/*.yaml; do
+        if [ -f "$yaml" ]; then
+          to_apply+=("$yaml")
+        fi
+      done
```
You're right that not all pipelines follow the naming convention :disappointed: 

But we do have the same kustomization checks as we do for tasks - the generated file should always be committed to the repo and up to date, so it shouldn't be necessary to re-run kustomize. What do you think about doing it this way:

for each YAML file in directory, if `.kind == "Pipeline"`, validate the file

kyngs replied:
Done

chmeliik reviewed (commented):
Awesome! One more suggestion, otherwise LGTM (and you can remove the demo commit now, thanks :+1: )
